forte/train_preprocessor.py

```diff
@@ -0,0 +1,355 @@
+        }
+
+    def _validate_config(self):
+        # Placeholder
+        pass
+
+    def _parse_request(self, request: Dict):
+        """
+        This method has two responsibilities:
+        1. parse the given data request and stored it internally
+        2. validate if the given data request is valid
+        """
+
+        assert "scope" in request, \
+            "Field not found for data request: `scope`"
+        assert "schemes" in request, \
+            "Field not found for data request: `schemes`"
+
+        resource_schemes: Dict[str, Dict] = {}
+        # Used for check dependency between different extractors
+        scheme_group: Dict[str, Dict] = {
+            "dependent": {}, "dependee": {}
+        }
+
+        for tag, scheme in request["schemes"].items():
+            assert "extractor" in scheme, \
+                "Field not found for data request scheme: `extractor`"
+            assert "type" in scheme, \
+                "Field not found for data request scheme: `type`"
+            resource_schemes[tag] = {}
+
+            if not isinstance(scheme["extractor"], BaseExtractor):
+                raise RuntimeError("Invalid extractor: ", scheme["extractor"])
+
+            extractor: BaseExtractor = scheme["extractor"]
+
+            # Track dependency
+            if hasattr(extractor, "based_on"):
+                if extractor.entry_type not in scheme_group["dependent"]:
+                    scheme_group["dependent"][extractor.entry_type] = set()
+                scheme_group["dependent"][extractor.entry_type].add(
+                    extractor)
+            else:
+                if extractor.entry_type not in scheme_group["dependee"]:
+                    scheme_group["dependee"][extractor.entry_type] = set()
+                scheme_group["dependee"][extractor.entry_type].add(
+                    extractor)
+
+            # Create default converter if there is no given converter
+            if "converter" not in scheme:
+                converter: Converter = Converter({})
+                scheme["converter"] = converter
+
+        # Check dependency
+        for _, dependent_extractors in scheme_group["dependent"].items():
+            for dependent_extractor in dependent_extractors:
+                based_on: Entry = dependent_extractor.based_on
+                if based_on not in scheme_group["dependee"]:
+                    raise ValueError(
+                        "Extractor {} needs the entry {} to do extraction "
+                        "processing but it is not extracted by any other "
+                        "extractors given in request".
+                            format(based_on, dependent_extractor.tag))
+
+        self._request = request
+        self._request_ready = True
+
+    def _build_vocab(self):
+        scope: EntryType = self._request["scope"]
+        schemes: Dict = self._request["schemes"]
+
+        # TODO: clear vocab?
```

Comment:
Will this todo be done? What happen if it is not done?

Reply:
Currently in our examples, there is no hurt if this is not done. It only hurts when someone uses the train preprocessor more than once.

Reply:
Maybe document this as an issue?

Reply:
Sure